### PR TITLE
feat: patch missing inner thoughts on new openai models

### LINKED ISSUE #1555: calling the/app/agents/id/messages interface never returns a result, suspected to be stuck in a dead loop
**Describe the bug**
When using the OpenAI GPT-4o model with MemGPT, calling the/app/agents/id/messages interface never returns a result, suspected to be stuck in a dead loop.

**Please describe your setup**
- [ x ] How did you install memgpt?
  - git clone
- [ x ] Describe your setup
  -  OS MacOS
  - Terminal

**Screenshots**

**Additional context**

OpenAI API ：{"error":{"message":"Authentication Error, ExceededTokenBudget: Current spend for token: 21.393555839999998; Max Budget for Token: 20.0","type":"auth_error","param":"None","code":401}}

/api/agents/4519a4c1-43a7-489f-9a8a-f00babbd5164/messages
Authorization: Bearer 123456
Content-Type: application/json

{
  "message": "xxx",
  "role": "user",
  "stream": false,
  "timestamp": "2024-07-17T14:42:14+08:00"
}

**MemGPT Config**

[defaults]
preset = memgpt_chat
persona = sam_pov
human = basic

[model]
model = gpt-4o
model_endpoint = https://ai-gateway-test.bondee-inc.com/v1
model_endpoint_type = openai
model_wrapper = null
context_window = 8192

[embedding]
embedding_endpoint_type = openai
embedding_endpoint = https://ai-gateway-test.bondee-inc.com/v1
embedding_model = text-embedding-ada-002
embedding_dim = 1536
embedding_chunk_size = 300

[archival_storage]
type = postgres
path = /root/.memgpt/chroma
uri = postgresql+pg8000://memgpt:memgpt@localhost:5432/memgpt

[recall_storage]
type = postgres
path = /root/.memgpt
uri = postgresql+pg8000://memgpt:memgpt@localhost:5432/memgpt

[metadata_storage]
type = postgres
path = /root/.memgpt
uri = postgresql+pg8000://memgpt:memgpt@localhost:5432/memgpt

[client]
anon_clientid = 00000000-0000-0000-0000-000000000000

---

If you're not using OpenAI, please provide additional information on your local LLM setup:

**Local LLM details**

If you are trying to run MemGPT with local LLMs, please provide the following information:

- [ ] The exact model you're trying to use (e.g. `dolphin-2.1-mistral-7b.Q6_K.gguf`)
- [ ] The local LLM backend you are using (web UI? LM Studio?)
- [ ] Your hardware for the local LLM backend (local computer? operating system? remote RunPod?)

### PR DESCRIPTION
**Pull request by cpacker**
TODOs:

Refactor code to convert to inner thought kwarg format

- [x] ~Add utility method inside of data types to convert `Message` objects~
  - See note*
  - Add utility to convert `ToolCall`s into versions that have a kwarg included (`add_inner_thoughts_to_tool_call`)
  - Add flag `put_inner_thoughts_in_kwargs` to `Message.to_openai_dict` which will call utility to make put `self.text` (the `content` / inner thoughts) into `ToolCall`s via the utility
- [x] Add utility method to convert functions list / tool objects (~will require dumping and reading json strings of arguments~)
  - `add_inner_thoughts_to_functions`
- [x] Add utility method to convert response from OpenAI response to strip inner thoughts from kwargs and into the content field for data storage
  - `unpack_inner_thoughts_from_kwargs`
- [x] Add flag to CLI to allow turning this on manually
  - I added a `--no-content` flag (`yes` = put inner thoughts in kwargs, `no` = leave as normal (gpt4 style), `default` = let us decide, eg for gpt4o turn it on, for gpt4 turn it off)
- [ ] TODO future PR - add support for our other API backends

*note: I was going to create a `Message -> Message` utility function that puts `self.text` into `self.tool_calls` (as an alternative to putting the logic for extracting inner thoughts from `self.text` into the function/tool call in the `to_*_dict()` methods), however upon further thought I think it's better to leave this logic inside the `_to_dict()` methods because it might vary significantly from provider to provider. Example of what I was going to do:
```python

# TODO have this use model_copy
def put_inner_thoughts_in_kwargs(message: Message) -> Message:
    """Convert a message that has inner monologue in self.text to one that has it inside a tool call"""
    # message_copy = message.model_copy()
    message_copy = copy.deepcopy(message)

    # TODO
    # step 1: assert the role is assistant and tool_calls > 1
    # step 2: copy self.text into self.tool_calls[0], and make self.text = None
    # step 3: return the message copy
```

**Please describe the purpose of this pull request.**

Fixes issues with missing `content` / inner thoughts on newer OpenAI models like 4o-mini

**How to test**

`memgpt run` with `gpt-4o-mini`

**Have you tested this PR?**

Yes, gpt4o-mini works fine now:
image

**Related issues or PRs**

Closes #1555

### CODE REVIEW

File: memgpt/models/pydantic_models.py [hunk 1]
```diff
@@ -13,6 +13,14 @@
 from memgpt.utils import get_human_text, get_persona_text, get_utc_time
 
 
+class OptionState(str, Enum):
```
Comment (cpacker) [inline memgpt/models/pydantic_models.py:16]: @sarahwooders I added a new type here for kwargs that are "optional booleans", basically yes, no, or "I don't care"/default

Comment (cpacker) [reply memgpt/models/pydantic_models.py:16]: I figured there's probably some standardized way for doing this or standardized type in `typing` (not just `Optional[bool]`) or Pydantic that I could use, but I couldn't find anything

File: memgpt/cli/cli.py [hunk 2]
```diff
@@ -410,6 +411,10 @@ def run(
     yes: Annotated[bool, typer.Option("-y", help="Skip confirmation prompt and use defaults")] = False,
     # streaming
     stream: Annotated[bool, typer.Option(help="Enables message streaming in the CLI (if the backend supports it)")] = False,
+    # whether or not to put the inner thoughts inside the function args
+    no_content: Annotated[
+        OptionState, typer.Option(help="Set to 'yes' for LLM APIs that omit the `content` field during tool calling")
```
Comment (sarahwooders) [inline memgpt/cli/cli.py:416]: I'm a bit confused by this description of the field - why is omitting `content` field during tool calling" related to putting inner thoughts inside the function args?

**Review by sarahwooders — APPROVED** [review]
lgtm!

File: memgpt/llm_api/llm_api_tools.py [hunk 3]
```diff
@@ -166,24 +259,47 @@ def create(
         printd("unsetting function_call because functions is None")
         function_call = None
 
+    # print("HELLO")
+
     # openai
     if llm_config.model_endpoint_type == "openai":
+
+        if inner_thoughts_in_kwargs == OptionState.DEFAULT:
```
Comment (cpacker) [inline memgpt/llm_api/llm_api_tools.py:267]: @sarahwooders here's an example where we're choosing for the user

**Pull request merged**